### PR TITLE
Feature/extended attachment

### PR DESCRIPTION
**Pull request by EvgeniyMurenko**

### CODE REVIEW

File: src/main/java/com/icthh/xm/tmf/ms/communication/messaging/handler/TemplatedEmailMessageHandler.java [hunk 1]
```diff
@@ -1,24 +1,34 @@
...
+import org.springframework.core.io.InputStreamSource;
+import org.springframework.stereotype.Service;
+
 import java.util.List;
 import java.util.Locale;
 import java.util.Map;
+import java.util.Objects;
 import java.util.stream.Collectors;
-import lombok.RequiredArgsConstructor;
-import lombok.extern.slf4j.Slf4j;
```
Comment (vkirichenko) [inline src/main/java/com/icthh/xm/tmf/ms/communication/messaging/handler/TemplatedEmailMessageHandler.java:20]: do not change order of imports

File: src/main/java/com/icthh/xm/tmf/ms/communication/messaging/handler/TemplatedEmailMessageHandler.java [hunk 2]
```diff
@@ -1,24 +1,34 @@
 package com.icthh.xm.tmf.ms.communication.messaging.handler;
 
-import static java.util.stream.Collectors.toMap;
```
Comment (vkirichenko) [inline src/main/java/com/icthh/xm/tmf/ms/communication/messaging/handler/TemplatedEmailMessageHandler.java:3]: do not change order of imports

File: src/main/resources/swagger/api.yml [hunk 3]
```diff
@@ -710,6 +711,10 @@ definitions:
       '@baseType':
         type: string
         description: The base type for use in polymorphic collections
+      attachmentType:
+          type: string
+    required:
+        - attachmentType
```
Comment (vkirichenko) [inline src/main/resources/swagger/api.yml:717]: old logic without this field should work

File: src/test/java/com/icthh/xm/tmf/ms/communication/messaging/handler/TemplatedEmailMessageHandlerTest.java [hunk 4]
```diff
@@ -1,27 +1,39 @@
 package com.icthh.xm.tmf.ms.communication.messaging.handler;
 
-import static org.assertj.core.api.Assertions.assertThat;
-import static org.mockito.ArgumentMatchers.any;
```
Comment (vkirichenko) [inline src/test/java/com/icthh/xm/tmf/ms/communication/messaging/handler/TemplatedEmailMessageHandlerTest.java:4]: do not change order of the imports

File: src/test/java/com/icthh/xm/tmf/ms/communication/messaging/handler/TemplatedEmailMessageHandlerTest.java [hunk 5]
```diff
@@ -59,4 +74,67 @@ public void testHandleMessage_shouldInvokeMailService() {
         List<String> allValues = captor.getAllValues();
         assertThat(allValues).containsExactly("content", "subject", "email", "sender");
     }
+
+    @Test
+    public void testHandleMessage_shouldSendEmailFromTemplateWithAttachments() {
```
Comment (vkirichenko) [inline src/test/java/com/icthh/xm/tmf/ms/communication/messaging/handler/TemplatedEmailMessageHandlerTest.java:79]: do not use _ in test methods name

**Review by vkirichenko — CHANGES_REQUESTED** [review]

File: src/main/resources/swagger/api.yml [hunk 6]
```diff
@@ -670,6 +670,7 @@ definitions:
   Attachment:
     type: object
     description: 'Complements the description of an element (for instance a product) through video, pictures...'
+    discriminator: attachmentType
```
Comment (vkirichenko) [inline src/main/resources/swagger/api.yml:673]: this is the only way how to achieve polymorphism in openapi

https://swagger.io/docs/specification/data-models/inheritance-and-polymorphism/

File: src/main/java/com/icthh/xm/tmf/ms/communication/messaging/handler/TemplatedEmailMessageHandler.java [hunk 7]
```diff
@@ -54,15 +65,42 @@ public CommunicationMessage handle(CommunicationMessageCreate messageCreate) {
...
-                locale,
-                templateName,
-                subject,
-                email,
-                sender,
-                objectModel
-            );
+            if (isExtendedAttachment) {
+                mailService.sendEmailFromTemplateWithAttachments(
+                    TenantContextUtils.getRequiredTenantKey(tenantContextHolder.getContext()),
```
[the remark this answers is not in the document]
Comment (EvgeniyMurenko) [reply src/main/java/com/icthh/xm/tmf/ms/communication/messaging/handler/TemplatedEmailMessageHandler.java:84]: +

File: src/main/java/com/icthh/xm/tmf/ms/communication/service/mail/MailService.java [hunk 8]
```diff
@@ -391,6 +391,7 @@ private String tokenizeKey(String key) {
     }
 
     private Optional<String> getI18nName(Map<String, String> name) {
+        if (name == null) return Optional.empty();
```
[the remark this answers is not in the document]
Comment (EvgeniyMurenko) [reply src/main/java/com/icthh/xm/tmf/ms/communication/service/mail/MailService.java:394]: +

File: src/test/java/com/icthh/xm/tmf/ms/communication/messaging/handler/TemplatedEmailMessageHandlerTest.java (hunk 4 above)
Comment (EvgeniyMurenko) [reply src/test/java/com/icthh/xm/tmf/ms/communication/messaging/handler/TemplatedEmailMessageHandlerTest.java:4]: +

File: src/test/java/com/icthh/xm/tmf/ms/communication/messaging/handler/TemplatedEmailMessageHandlerTest.java (hunk 5 above)
Comment (EvgeniyMurenko) [reply src/test/java/com/icthh/xm/tmf/ms/communication/messaging/handler/TemplatedEmailMessageHandlerTest.java:79]: +

File: src/main/java/com/icthh/xm/tmf/ms/communication/messaging/handler/TemplatedEmailMessageHandler.java [hunk 9]
```diff
@@ -55,18 +66,44 @@ public CommunicationMessage handle(CommunicationMessageCreate messageCreate) {
...
+            convertAttachments(attachments)
+        );
+    }
+
+    private Map<String, InputStreamSource> convertAttachments(List<Attachment> attachments) {
+        if (attachments == null || attachments.isEmpty()) {
+            return Map.of();
+        }
+
+        if (attachments.get(0) instanceof ExtendedAttachment) {
```
Comment (vkirichenko) [inline src/main/java/com/icthh/xm/tmf/ms/communication/messaging/handler/TemplatedEmailMessageHandler.java:94]: let's discuss general solutution without attachments.get(0)

Comment (EvgeniyMurenko) [reply src/main/java/com/icthh/xm/tmf/ms/communication/messaging/handler/TemplatedEmailMessageHandler.java]: +

**Review by vkirichenko — APPROVED** [review]

File: src/main/resources/swagger/api-extension.yml [hunk 10]
```diff
@@ -1,44 +1,67 @@
 swagger: '2.0'
+info:
+    title: Communication Management Extentions
+    version: 1.0.0
+paths:
+    /ping:
```
Comment (amedvedchuk) [inline src/main/resources/swagger/api-extension.yml:6]: ping? for what?

File: src/main/resources/swagger/api.yml (hunk 6 above)
Comment (amedvedchuk) [reply src/main/resources/swagger/api.yml:673]: @vkirichenko why do not use `@type` field as discriminator?
```
'@type':
        type: string
        description: The class type of the actual resource (for type extension).
```

seems this is it designed for.

Comment (vkirichenko) [reply src/main/resources/swagger/api.yml:673]: @EvgeniyMurenko let's check and discuss this

File: src/main/resources/swagger/api-extension.yml (hunk 10 above)
Comment (vkirichenko) [reply src/main/resources/swagger/api-extension.yml:6]: let's just remove this

File: src/main/resources/swagger/api.yml (hunk 6 above)
Comment (EvgeniyMurenko) [reply src/main/resources/swagger/api.yml]: @amedvedchuk done

File: src/main/resources/swagger/api-extension.yml (hunk 10 above)
Comment (EvgeniyMurenko) [reply src/main/resources/swagger/api-extension.yml]: @amedvedchuk done

**Review by amedvedchuk — APPROVED** [review]

**Pull request merged**